node-api/src/error.rs

```diff
@@ -115,44 +136,84 @@ impl DispatchError {
 			},
 		};
 
-		let module_variant_idx =
-			variant.variants.iter().find(|v| v.name == "Module").map(|v| v.index);
-		let module_variant_idx = match module_variant_idx {
-			Some(idx) => idx,
+		let variant_name =
+			variant.variants.iter().find(|v| v.index == bytes[0]).map(|v| v.name.as_str());
+		let name = match variant_name {
+			Some(name) => name,
 			None => {
-				warn!("Can't decode error: sp_runtime::DispatchError does not have a 'Module' variant");
+				warn!("Can't decode error: sp_runtime::DispatchError does not have a name variant");
 				return DispatchError::Other(bytes.into_owned())
 			},
 		};
+		// The remaining bytes are the specific error to decode:
+		let mut specific_bytes = &bytes[1..];
 
-		// If the error bytes don't correspond to a ModuleError, just return the bytes.
-		// This is perfectly reasonable and expected, so no logging.
-		if bytes[0] != module_variant_idx {
-			return DispatchError::Other(bytes.into_owned())
+		match name {
+			"Module" => Self::decode_module_error(specific_bytes, metadata), // We apply custom logic to transform the module error into the outward facing version
+			"Token" => {
+				let token_error = match TokenError::decode(&mut specific_bytes) {
+					Ok(err) => err,
+					Err(_) => {
+						warn!("Can't decode token error: TokenError does not match known formats");
+						return DispatchError::Other(bytes.to_vec())
+					},
+				};
+				DispatchError::Token(token_error)
+			},
+			"Arithmetic" => {
+				let arithmetic_error = match ArithmeticError::decode(&mut specific_bytes) {
+					Ok(err) => err,
+					Err(_) => {
+						warn!("Can't decode token error: ArithmeticError does not match known formats");
+						return DispatchError::Other(bytes.to_vec())
+					},
+				};
+				DispatchError::Arithmetic(arithmetic_error)
+			},
+			"-" => {
```

Comment:
Do you know why this is "-" as name and not `Transactional`? It seems strange to me.

Reply:
My bad